Add draft of ISOLATE_STDLIB_MODULES.rst (#24)
encukou/abi3: ISOLATE_STDLIB_MODULES.rst

PEP: XXX
Title: Isolating modules in the standard library
Author: Erlend Egeberg Aasland <erlend.aasland@protonmail.com>, Petr Viktorin <encukou@gmail.com>
Discussions-To: XXX
Status: Draft
Type: Standards
Requires: 489, 573, 630
Created: 16-Feb-2022
Python-Version: 3.11
Post-History: XXX

Abstract
========

Extensions in the standard library will be converted to multi-phase
initialization (:pep:`489`) and where possible, all state will be
stored on module objects rather than in process-global variables.

Note on Backdating
==================

> Much of this proposal has already been implemented, partly to solve related
> issues and partly because of misunderstandings and lack of coordination.
> 

Reviewer: One last nitpick: IMO, the important piece of information for the reader is that much of this has already been implemented. There's probably a lot of reasons why (finding all of them would be a time-consuming task), but I'm not sure those reasons are valuable information in this context. Suggesting a simplification:

```suggestion
Much of this proposal has already been implemented.
```

For example, deleting such a module from ``sys.modules`` and re-importing it
will result in a fresh module instance, isolated from any previously loaded
instances.

This may affect code that expected the previous behavior: globals of
extension modules were shallowly copied from the first loaded module.

Security Implications
=====================

None known.

How to Teach This
=================

A large part of this proposal is a HOWTO aimed at experienced users,
which will be moved to the documentation.

Beginners should not be affected.

Reference Implementation
========================

Main branch, particularly commits for these issues:

- https://bugs.python.org/issue40077
- https://bugs.python.org/issue46417
- https://bugs.python.org/issue1635741

Rejected Ideas
==============

Keep things as they were and not break anything
-----------------------------------------------

XXX Someone write something here please

Open Issues
===========

See XXX above.

Types that do not need module state access
------------------------------------------

If Python gets a per-interpreter GIL, these types will need to either

- be converted to heap types, or

> - become immortal (see :pep:`683`).

Author: This is covered in the pre-interpreter GIL PEP, so I'll remove it.

PEP: XXX
Title: Isolating modules in the standard library
Author: Petr Viktorin <encukou@gmail.com>, Erlend Egeberg Aasland <erlend.aasland@protonmail.com>
Discussions-To: XXX
Status: Draft
Type: Standards
Requires: 489, 573, 630
Created: 16-Feb-2022
Python-Version: 3.11
Post-History: XXX

Abstract
========

Extensions in the standard library will be converted to multi-phase
initialization (:pep:`489`) and where possible, all state will be
stored on module objects rather than in process-global variables.

Note on Backdating
==================

Much of this proposal has already been implemented, partly to solve related
issues and partly because of misunderstandings and lack of coordination.

We submit this PEP to explain the changes, seek consensus on
whether they are good,  propose changing modules that do not exhibit

> issues, and set best practices for new modules.

Reviewer: _[...] propose changing modules that do not exhibit issues [...]_

This reads a little bit weird; did you mean "propose _not_ changing modules that do not exhibit issues"? In that case, we should consider rewording it.

We are aware that the pre-existing changes might need to be reverted if issues

> are found.

Reviewer: The first heap types were added in Python 3.2. Some of these changes are _very_ old. If one concludes there is no gain in converting new types (net win = 0), then there is also no gain in reverting very old changes; IMO it would just be needless code churn, which would be very counter productive (risk of introducing new regressions, require a _lot_ of reviewer resources).

There is already a section further down that says if issues are found, they must be resolved. That should suffice. I suggest dropping this sentence.

Reviewer: > There is already a section further down that says if issues are found, they must be resolved.

Issues could be general (the overall process) or specific (for example module or type specific). One way of resolving a module specific issue could be to revert changes for that specific module; that can be resolved on a bug ticket; we don't need it in the PEP.

The PEP takes the form of the proposal that should have been submitted,

> talking about the situation before the proposed changes as current.

Reviewer: We should consider removing this sentence or rewording it so it is easier to grasp.

Author: Do you have a suggestion?

Reviewer: I'll try to come up with something.

Reviewer: Re-reading this, I do not see what new value or information it adds. The two first sentences under this sub-heading already explains what the reader needs to know. Here is my suggestion:

```suggestion
```

generally to any extension module (not just the standard library).

It is an integral part of this proposal. Read it first.

This PEP discusses specifics of the standard library.

Specification
=============

The body of :pep:`630` will be converted to a HOWTO in the Python
documentation, and PEP will be retired (marked Final).

All extension modules in the standard library will be converted to multi-phase
initialization introduced in :pep:`484`.

All stdlib extension modules will be *isolated*. That is:

- Types, functions and other objects defined by the module will either be
  immutable, or not shared with other module instances.

- State specific to the module will not be shared with other module instances,
  unless it represents global state.
  For example, ``_csv.field_size_limit`` will get/set a module-specific
  number. On the other hand, functions like ``readline.get_history_item`` or
  ``os.getpid`` will continue to work with state that is process-global
  (external to the module, and possibly shared across other libraries, including
  non-Python ones).

Conversion to heap types
------------------------

> Types whose methods need access to “their” module instance will be converted

Reviewer: ```suggestion
Types whose methods need access to their module instance will be converted
```

to heap types following :pep:`630`, with the following considerations:

- All heap types **must fully implement the GC protocol**. See
  [bpo-42972](https://bugs.python.org/issue42972).
  (XXX move this to PEP 630)

- All standard library types that used to be static types should remain
  immutable. Heap types must be defined with the `Py_TPFLAGS_IMMUTABLE_TYPE`
  flag to retain immutability.
  See `bpo-43908 <https://bugs.python.org/issue43908>`__.

> - A static type with ``tp_new = NULL`` does not have public constructor, but

Reviewer: ```suggestion
- A static type with ``tp_new = NULL`` does not have a public constructor, but
```

  heap types inherit the constructor from the base class. Make sure types that
  previously were impossible to instantiate retain that feature; use
  `Py_TPFLAGS_DISALLOW_INSTANTIATION`. Add tests using
  `test.support.check_disallow_instantiation()`. See
  `bpo-43916 <https://bugs.python.org/issue43916>`__.

- Converted heap types may unintentionally become serializable
  (``pickle``-able). Test that calling ``pickle.dumps`` has the same result
  before and after conversion, and if the test fails, add a ``__reduce__``
  method that raises ``TypeError``. See `PR-21002 <https://github.com/python/cpython/pull/21002/files>`__
  for an example.

- Use strong "back-refs" to the module object to ensure the module state
  pointer never outlives objects that access module state. Keep this in mind
  for external library callbacks that access module state.
  (XXX move this to PEP 630)

These issues will be added to the Devguide to help any future conversions.

> If another kind of issue is found, the module in questions should be unchanged

Reviewer: ```suggestion
If another kind of issue is found, the module in question should be unchanged
```

until a solution is found and added to the Devguide, and already
converted modules are checked and fixed.

> Static types that do not need module state access and have no other reason to
> be converted should stay static.

Reviewer: ```suggestion
Static types that do not need module state access, and have no other reason to
be converted, should stay static.
```

Testing
-------

> XXX How should this be tested?

Reviewer: We've already described how to test for disallowed instantiation and serialisation, so how about moving those advices down here? We should also include a short example of how to test for immutability (try to create a new attribute on the type; make sure `TypeError` is raised).

Author: Right. It makes sense to have the testing info in the list of considerations, instead of a new section.

Process
-------

> The following process should be added to the Devguide, and stay there until
> all modules are converted.

Reviewer: ```suggestion
The following process should be added to the Devguide, and remain until
all modules are converted.
```

Any new findings should be documented there or in the general HOWTO.

Part 1: Preparation
...................

1. Open a discussion, either on the bug tracker or on Discourse. Involve the
   module maintainer and/or code owner. Explain the reason and rationale for
   the changes.

> 2. Identify global state performance bottlenecks, if there are such. Create a
>    proof-of-concept implementation and measure the performance impact. `pyperf`
>    is a good tool for benchmarking.

Reviewer: ```suggestion
2. Identify global state performance bottlenecks.
   Create a proof-of-concept implementation, and measure the performance impact.
   `pyperf` is a good tool for benchmarking.
```

3. Create an implementation plan. For small modules with few types, a single PR
   may do the job. For larger modules with lots of types, and possibly also
   external library callbacks, multiple PR's will be needed.

Part 2: Implementation
......................

Note: this is a suggested implementation plan for a complex module, based on
lessons learned with other modules. Feel free to simplify it for
smaller modules.

1. Add Argument Clinic where possible; it enables you to easily use the
   defining class to fetch module state from type methods.
2. Prepare for module state; establish a module state `struct`, add an instance
   as a static global variable, and create helper stubs for fetching the module
   state.

> 3. Add relevant global variables to the module state `struct` and modify code

Reviewer: ```suggestion
3. Add relevant global variables to the module state `struct`, and modify code
```

   that accesses the global state to use the module state helpers instead. This
   step may be broken into several PR's.
4. Where necessary, convert heap types to static types.
5. Convert the global module state struct to true module state.
6. Implement multi-phase initialisation.

> Preferably, steps 4 through 6 should all land early in a single alpha
> development phase.

Reviewer: ```suggestion
Steps 4 through 6 should preferably land in a single alpha development phase.
```

Backwards Compatibility
=======================

Extension modules in the standard library will now be loadable more than once.
For example, deleting such a module from ``sys.modules`` and re-importing it
will result in a fresh module instance, isolated from any previously loaded
instances.

This may affect code that expected the previous behavior: globals of
extension modules were shallowly copied from the first loaded module.

Security Implications
=====================

None known.

How to Teach This
=================

A large part of this proposal is a HOWTO aimed at experienced users,
which will be moved to the documentation.

Beginners should not be affected.

Reference Implementation
========================

Main branch, particularly commits for these issues:

- https://bugs.python.org/issue40077
- https://bugs.python.org/issue46417

> - https://bugs.python.org/issue1635741

Reviewer: I think we should add links to the actual implementation instead (and maybe one or two reference PRs); asking someone to extract relevant bits and pieces from bpo-1635741 is ... not nice :)

Reviewer: The `csv` module should be a good candidate.

Rejected Ideas
==============

> Keep things as they were and not break anything

Reviewer: Would not this section make more sense in the subinterpreters or per-interpreter GIL PEPs? It feels kinda out of place here. Also, if we are to keep it, I would rephrase it to something more accurate and direct, such as _"Do not isolate modules in the stdlib"_.

It is not clear to me what "things" is. I would guess that the sentence implies _"don't convert static types to heat types"_, _"don't convert global state to module state"_, _"don't apply PEP 630 to the stdlib"_, or maybe _"drop subinterpreter support"_. It is hard to tell. Currently, I'm afraid this will only be a source of confusion.

Author: Yeah, the title should have XXX as well. I guess “Do not isolate modules in the stdlib” works best, because it needs the other changes.

Reviewer: We should add a short sentence about why having stdlib extension modules _without_ multi-init support and _with_ global state is a problem; for example, such modules would hamper a lot of Eric's work-in-progress.

which will be moved to the documentation.

Beginners should not be affected.

Reference Implementation
========================

Most of the changes are now in the main branch, as commits for these issues:

- `bpo-40077, Convert static types to heap types: use PyType_FromSpec() <https://bugs.python.org/issue40077>`_
- `bpo-40077, Clear static types in Py_Finalize() for embedded Python <https://bugs.python.org/issue46417>`_
- `bpo-1635741, Py_Finalize() doesn't clear all Python objects at exit <https://bugs.python.org/issue1635741>`_

As an example, changes and fix-ups done in the `_csv` module are:

- `GH-23224, Remove static state from the _csv module <https://github.com/python/cpython/pull/23224>`_
- `GH-26008, Allow subclassing of ``csv.Error`` <https://github.com/python/cpython/pull/26008>`_
- `GH-26074, Add GC support to _csv heap types <https://github.com/python/cpython/pull/26074>`_
- `GH-26351, Make heap types converted during 3.10 alpha immutable <https://github.com/python/cpython/pull/26351>`_

Rejected Ideas
==============

Do not isolate modules in the standard library
----------------------------------------------

> XXX Someone write something here please

Reviewer: IMO we can reference Eric's work-in-progress (per-interpreter GIL), subinterpreters issues (multiple subinterpreters + static global state => 💥), ... anything else?

Author: Do we want to make this PEP depend on Eric's PEP?
IMO it would be better if it can stand on its own.

Reviewer: > IMO it would be better if it can stand on its own.

Yes, that's preferable.

Author: So, let's remove the Rejected Ideas section altogether and let the discussion generate some.

Are you happy with your name on this? I think it's ready to post to python-dev.

Reviewer: Sounds good to me!

Open Issues
===========

> See XXX above.

Reviewer: This is already covered in the heap types section, right? I guess we could remove this then.

Reference Implementation
========================

Most of the changes are now in the main branch, as commits for these issues:

- `bpo-40077, Convert static types to heap types: use PyType_FromSpec() <https://bugs.python.org/issue40077>`_

> - `bpo-40077, Clear static types in Py_Finalize() for embedded Python <https://bugs.python.org/issue46417>`_

Reviewer: ```suggestion
- `bpo-46417, Clear static types in Py_Finalize() for embedded Python <https://bugs.python.org/issue46417>`_
```
